pkg/extract/types/data/detection/condition/criteria/criterion/criterion.go

```diff
@@ -2,162 +2,200 @@ package criterion
-			case x.FixStatus == nil && y.FixStatus == nil:
-				return 0
-			case x.FixStatus == nil && y.FixStatus != nil:
-				return -1
-			case x.FixStatus != nil && y.FixStatus == nil:
-				return +1
-			default:
-				return fixstatusTypes.Compare(*x.FixStatus, *y.FixStatus)
-			}
-		}(),
-		packageTypes.Compare(x.Package, y.Package),
-		func() int {
-			switch {
-			case x.Affected == nil && y.Affected == nil:
-				return 0
-			case x.Affected == nil && y.Affected != nil:
-				return -1
-			case x.Affected != nil && y.Affected == nil:
-				return +1
-			default:
-				return affectedTypes.Compare(*x.Affected, *y.Affected)
-			}
-		}(),
 	)
 }
 
 type Query struct {
-	Package *QueryPackage
-	CPE     *string
-}
-
-type QueryPackage struct {
-	Name       string
-	Version    string
-	SrcName    string
-	SrcVersion string
-	Arch       string
-	Repository string
-	Functions  []string
+	Version      []vcTypes.Query
+	NotExistence *necTypes.Query
 }
 
-func (c Criterion) Accept(ecosystem ecosystemTypes.Ecosystem, query Query) (bool, error) {
-	if !c.Vulnerable {
-		return false, nil
-	}
-
-	switch {
-	case query.Package != nil:
-		name, version, arch := query.Package.Name, query.Package.Version, query.Package.Arch
-		if slices.Contains(c.Package.Architectures, "src") {
-			name, version, arch = query.Package.SrcName, query.Package.SrcVersion, "src"
+func (c Criterion) Contains(query Query) (bool, error) {
+	switch c.Type {
+	case CriterionTypeVersion:
+		if len(query.Version) == 0 {
+			return false, errors.New("query is not set for version criterion")
 		}
-
-		isAccept, err := c.Package.Accept(packageTypes.Query{Package: &packageTypes.QueryPackage{
-			Name:       name,
-			Arch:       arch,
-			Repository: query.Package.Repository,
-			Functions:  query.Package.Functions,
-		}})
-		if err != nil {
-			return false, errors.Wrap(err, "package accept")
-		}
-
-		if !isAccept {
-			return false, nil
+		for _, q := range query.Version {
+			isAccept, err := c.Version.Accept(q)
+			if err != nil {
+				return false, errors.Wrap(err, "version criterion accept")
+			}
+			if isAccept {
+				return true, nil
+			}
 		}
-
-		if c.Affected == nil {
-			return true, nil
+		return false, nil
+	case CriterionTypeNotExistence:
+		if query.NotExistence == nil {
+			return false, errors.New("query is not set for not existence criterion")
 		}
-
-		isAccept, err = c.Affected.Accept(ecosystem, version)
+		isAccept, err := c.NotExistence.Accept(*query.NotExistence)
 		if err != nil {
-			return false, errors.Wrap(err, "affected accept")
+			return false, errors.Wrap(err, "not existence criterion accept")
 		}
-
 		return isAccept, nil
-	case query.CPE != nil:
-		isAccept, err := c.Package.Accept(packageTypes.Query{CPE: query.CPE})
-		if err != nil {
-			return false, errors.Wrap(err, "package accept")
-		}
+	default:
+		return false, errors.Errorf("unexpected criterion type. expected: %q, actual: %q", []CriterionType{CriterionTypeVersion, CriterionTypeNotExistence}, c.Type)
```

Comment:
int type values are not very nice for error message, I guess.

Reply:
If `String()` is provided, the corresponding string will be displayed instead of int type value.
https://github.com/MaineK00n/vuls-data-update/pull/269/files#diff-eb895a45da29219fa72841a17c220501790fb32bc22dc9cbf6c8493c3341ec00R24-R33

https://go.dev/play/p/d09kmGy-4CF